include/real/real_explicit.hpp

```diff
@@ -221,102 +222,82 @@ namespace boost {
-                        dot_amount++;
-                    }
+                std::size_t dot_pos = number.find('.');
+                if (dot_pos == std::string::npos) {
+                    integer_part = number;
+                    decimal_part = "";
                 }
-
-                bool there_is_dot = dot_amount > 0;
-                int exponent = 0;
-                dot_amount = 0;
-
-                // Check whether the number is explicitly specified as positive or negative.
-                unsigned int first_index = 0;
-                if (number.at(first_index) == '+') {
+                else {
+                    integer_part = number.substr(0, dot_pos);
+                    if(has_exponent)
+                        decimal_part = number.substr(dot_pos+1, e_pos - dot_pos -1);
+                    else
+                        decimal_part = number.substr(dot_pos+1);
+                }
+                if(integer_part[0]=='+') {
                     this->_positive = true;
-                    first_index++;
-                } else if (number.at(first_index) == '-') {
+                    integer_part = integer_part.substr(1);
+                }
+                if(integer_part[0]=='-') {
                     this->_positive = false;
-                    first_index++;
+                    integer_part = integer_part.substr(1);
                 }
-
-                // Remove zeros from the lefts side
-                // Note: We know at this point that number.length > 0 because of the first check
-                unsigned int last_index = (unsigned int)number.length() - 1;
-                while (last_index > first_index && (number.at(last_index) == '0' || number.at(last_index) == '.')) {
-                    if (number.at(last_index) == '.') {
-                        dot_amount++;
-                    }
-                    last_index--;
-                    if (!there_is_dot || dot_amount > 0) {
-                        exponent++; // remove zeros from the integer part increases the exponent
+                int i = 0;
+                while (integer_part[i]=='0'&&i<integer_part.length()) {
+                    ++i;
+                }
+                integer_part = integer_part.substr(i);
+                i = decimal_part.length()-1;
+                while (decimal_part[i]=='0'&&i>0) {
+                    --i;
+                }
+                decimal_part = decimal_part.substr(0, i+1);
+                //decimal and integer parts stripped of zeroes
+                int exponent = integer_part.length() + add_exponent;
+                if(decimal_part.empty()) {
+                    i = integer_part.length()-1;
+                    while (integer_part[i]=='0'&&i>0)
+                        --i;
+                    integer_part = integer_part.substr(0, i+1);
+                }
+                if(integer_part.empty()) {
+                    i = 0;
+                    while (decimal_part[i]=='0'&&i<decimal_part.length()) {
```

Comment:
decimal_part[i] == '0' && i < decimal_part.length()
Put spaces between operators and operands.